### ISSUE TITLE
Arcface Paper vs Code Understanding

### ISSUE THREAD

**Issue by ElisonSherton** [KevinMusgrave/pytorch-metric-learning#537]
Hey guys! 

Good Day! Hope everyone's keeping well. I had one question regarding the computation of cosine of target logit in ArcFace loss.

In the source-code for [arcface loss](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/64f47bafd12a0f68a7cffa4c70146918de92eabf/src/pytorch_metric_learning/losses/large_margin_softmax_loss.py#L86
), I found the following
```python
  def modify_cosine_of_target_classes(self, cosine_of_target_classes):
      angles = self.get_angles(cosine_of_target_classes)
      return torch.cos(angles + self.margin)
```

There is an addition of margin to the angle and then we compute the cosine. While this is alright for all cases where the feature embedding makes an angle less than `180 deg - margin deg`, when our angle is higher, we would be in the third quadrant and the cos value becomes less negative i.e. increases there. So, we would want to penalize our model more for highly obtuse angles but when the angle becomes too large, our penalty starts to soften. Hence the authors of the insightface paper in their [repo `insightface`](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_torch/losses.py#L85) handle it as follows

```python
final_target_logit = torch.where(target_logit > self.theta, cos_theta_m, target_logit - self.sinmm)
```
This is to keep the loss function monotonically decreasing with respect to increasing angles between target center and the feature embedding.

However, despite this change, pml's arcface gives decent results; any ideas on what is going on here or did I miss any code where you have added this check already?

Thanks a lot,
Vinayak.

**Comment by KevinMusgrave** [comment]
This is an interesting point that I don't think I ever considered. So I'm marking it as a bug.

**Comment by ElisonSherton** [comment]
Thanks for the prompt action Kevin. Appreciate it. Would like to suggest a pull request in the future by implementing this in the library :)

**Comment by KevinMusgrave** [comment]
I'll keep this open until it's resolved

**Comment by KevinMusgrave** [comment]
Fixed in v1.7.0.

Relevant links:

- https://github.com/KevinMusgrave/pytorch-metric-learning/pull/539
- https://github.com/deepinsight/insightface/issues/2126
- https://github.com/ronghuaiyang/arcface-pytorch/issues/48
